insights-notification-schemas-java/src/test/java/com/redhat/cloud/notifications/ingress/TestSerialization.java

```diff
@@ -1,164 +1,342 @@
+                .withAccountId("account-id")
+                .withOrgId("my-org-id")
+                .withBundle("my-bundle")
+                .withApplication("my-app")
+                .withEventType("my-event-type")
+                .withTimestamp(LocalDateTime.now())
+                .withContext(
+                        new Context.ContextBuilder().withAdditionalProperty("foo", "bar").build()
+                )
+                .withRecipients(
+                        List.of(
+                                new Recipient.RecipientBuilder().build()
+                        )
+                )
+                .withEvents(
+                        List.of(
+                                new Event.EventBuilder()
+                                        .withMetadata(new Metadata.MetadataBuilder().build())
+                                        .withPayload(new Payload.PayloadBuilder()
+                                                .withAdditionalProperty("foo", "bar")
+                                                .build())
+                                        .build()
+                        )
+                )
+                .build();
+    }
 
-        // Decoded does not have "version" field
-        assertThrows(AvroRuntimeException.class, () -> decoded.get("version"));
+    private void testParserEncode(Action action, boolean fails) {
+        Executable encoding = () -> Parser.encode(action);
+        if (fails) {
+            assertThrows(ParsingException.class, encoding);
+        } else {
+            assertDoesNotThrow(encoding);
+        }
     }
+
+    private void testDate(String stringAsDate, LocalDateTime dateTime) {
+        String serialized = String.format(
+                "{\"bundle\":\"my-bundle\",\"application\":\"Policies\",\"event_type\":\"Any\",\"timestamp\":\"%s\",\"account_id\":\"testTenant\",\"context\":{},\"events\":[]}\n",
+                stringAsDate
+        );
+        assertEquals(
+                dateTime,
+                Parser.decode(serialized).getTimestamp()
+        );
+    }
+
+    private Optional<Integer> getInteger(String string) {
+        try {
+            return Optional.of(Integer.parseInt(string));
+        } catch (NumberFormatException nfe) {
+            return Optional.empty();
+        }
+    }
+
+    private boolean hasField(JsonNode node, String field) {
+        Optional<Integer> numericField = getInteger(field);
+        if (numericField.isPresent()) {
+            return node.has(numericField.get());
+        }
+
+        return node.has(field);
+    }
+
+    private JsonNode getField(JsonNode node, String field) {
+        Optional<Integer> numericField = getInteger(field);
+        if (numericField.isPresent()) {
+            return node.get(numericField.get());
+        }
+
+        return node.get(field);
+    }
+
+    private void testRequiredField(String field, boolean isRequired, String template) throws JsonProcessingException {
```

Comment:
```suggestion
    /**
     * This method removes a field from a valid template and checks one of these two cases:
     - if {@code isRequired} is true, the field removal should trigger an exception throw when the template is decoded
     - otherwise, the field removal should not cause any exception during the decoding as the field is expected to be optional 
     */
    private void testRequiredField(String field, boolean isRequired, String template) throws JsonProcessingException {
```